src/spec.c

```diff
@@ -319,124 +319,230 @@ static int parseTextField(FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
+// Tries to get vector data type from ac. This function need to stay updated with
+// the supported vector data types list of VecSim.
+static int parseVectorField_GetType(ArgsCursor *ac, VecSimType *type) {
   const char *typeStr;
   size_t len;
+  int rc;
   if ((rc = AC_GetString(ac, &typeStr, &len, 0)) != AC_OK) {
-    QERR_MKBADARGS_AC(status, "vecsim type", rc);
-    return 0;
+    return rc;
   }
   if (!strncasecmp(VECSIM_TYPE_FLOAT32, typeStr, len)) 
-    fs->vecSimParams.type = VecSimType_FLOAT32;
+    *type = VecSimType_FLOAT32;
   else if (!strncasecmp(VECSIM_TYPE_FLOAT64, typeStr, len)) 
-    fs->vecSimParams.type = VecSimType_FLOAT64;
+    *type = VecSimType_FLOAT64;
   else if (!strncasecmp(VECSIM_TYPE_INT32, typeStr, len)) 
-    fs->vecSimParams.type = VecSimType_INT32;
+    *type = VecSimType_INT32;
   else if (!strncasecmp(VECSIM_TYPE_INT64, typeStr, len)) 
-    fs->vecSimParams.type = VecSimType_INT64;
+    *type = VecSimType_INT64;
   else {
-    QERR_MKBADARGS_AC(status, "vecsim type", AC_ERR_PARSE);
-    return 0;
-  }
-
-  // parse size
-  long long size;
-  if ((rc = AC_GetSize(ac, &fs->vecSimParams.size, 0)) != AC_OK) {
-    QERR_MKBADARGS_AC(status, "vecsim size", rc);
-    return 0;
+    return AC_ERR_ENOENT;
   }
+  return AC_OK;
+}
 
-  // parse metric
+// Tries to get distance metric from ac. This function need to stay updated with
+// the supported distance metric functions list of VecSim.
+static int parseVectorField_GetMetric(ArgsCursor *ac, VecSimMetric *metric) {
   const char *metricStr;
+  size_t len;
+  int rc;
   if ((rc = AC_GetString(ac, &metricStr, &len, 0)) != AC_OK) {
-    QERR_MKBADARGS_AC(status, "vecsim metric", rc);
-    return 0;
+    return rc;
   }
   if (!strncasecmp(VECSIM_METRIC_IP, metricStr, len)) 
-    fs->vecSimParams.metric = VecSimMetric_IP;
+    *metric = VecSimMetric_IP;
   else if (!strncasecmp(VECSIM_METRIC_L2, metricStr, len)) 
-    fs->vecSimParams.metric = VecSimMetric_L2;
+    *metric = VecSimMetric_L2;
   else if (!strncasecmp(VECSIM_METRIC_COSINE, metricStr, len)) 
-    fs->vecSimParams.metric = VecSimMetric_Cosine;
+    *metric = VecSimMetric_Cosine;
   else {
-    QERR_MKBADARGS_AC(status, "vecsim metric", AC_ERR_PARSE);
+    return AC_ERR_ENOENT;
+  }
+  return AC_OK;
+}
+
+static int parseVectorField_hnsw(FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
+  int rc;
+
+  // HNSW mandatory params.
+  bool mandtype = false;
+  bool mandsize = false;
+  bool mandmetric = false;
+
+  // Get number of parameters
+  size_t expNumParam, numParam = 0;
+  if ((rc = AC_GetSize(ac, &expNumParam, 0)) != AC_OK) {
+    QERR_MKBADARGS_AC(status, "vecsim number of parameters", rc);
     return 0;
   }
 
-  // parse algorithm
-  VecSimAlgo algorithm;
-  const char *algStr;
-  if ((rc = AC_GetString(ac, &algStr, &len, 0)) != AC_OK) {
-    QERR_MKBADARGS_AC(status, "vecsim algorithm", rc);
+  while (expNumParam > numParam && !AC_IsAtEnd(ac)) {
+    if (AC_AdvanceIfMatch(ac, VECSIM_TYPE)) {
```

Comment:
In RediSearch we always use the number of parameters and not the number of pairs.
In the examples, we have `3` while it should be `6`.
@MeirShpilraien WDYT?

Reply:
Agree, we should be consistent.